pkg/controllers/manila/manila.go

```diff
@@ -123,6 +145,75 @@ func (c *ManilaController) sync(ctx context.Context, syncCtx factory.SyncContext
 	return c.setEnabledCondition(ctx)
 }
 
+func (c *ManilaController) getFsGroupPolicy(ctx context.Context) storagev1.FSGroupPolicy {
+
+	// Set the fsGroupPolicy param to the default value
+	fsGroupPolicy := "None"
+
+	// Try to get the current value defined for the operator
+	if c.isFSGroupPolicyAllowed(os.Getenv("CSI_FSGROUP_POLICY")) {
+		fsGroupPolicy = os.Getenv("CSI_FSGROUP_POLICY")
+	}
+
+	return storagev1.FSGroupPolicy(string(fsGroupPolicy))
+}
+
+func (c *ManilaController) applyCSIDriver(ctx context.Context, spec storagev1.CSIDriverSpec) error {
+	klog.V(4).Infof("Starting CSI driver config refresh")
+	defer klog.V(4).Infof("CSI driver config refresh finished")
+
+	var errs []error
+
+	crd, err := c.kubeClient.StorageV1().CSIDrivers().Get(ctx, "manila.csi.openstack.org", metav1.GetOptions{})
+
+	if err != nil {
+		crd = c.generateCSIDriver()
```

Comment:
@jsafrane in L167 I get the existing CSIDriver object and patch it with the generated spec.
If an error occurs it means the object is not there, and I generate it with the existing function.

This shouldn't happen because as per your previous comment/review we want the operator to apply static resources when it starts, and `CSIDriver` object is one of them.

However, if I keep `csidriver.yaml` among the applied static resources [1], after updating the `CSISpec` map the operator reconciles it with the original object defined in the static file, which means we can't take advantage of [2] because after patching the object the `StaticResourceController` deletes and recreates it.

If I remove [1] and manually apply `csidriver.yaml` (or if I rely on the `generateCSIDriver` function), it works as expected.

Do you have a suggestion on how we can remove this `generateCSIDriver` function and apply the static resource defined in the `csidriver.yaml` file without making it part of the resources reconciled by the `StaticResourceController`?

[1] https://github.com/openshift/csi-driver-manila-operator/blob/master/pkg/operator/starter.go#L77
[2] https://github.com/openshift/library-go/pull/1324/commits/c4d0b28d0d214f590dd499961b1a55c209ef1822

Reply:
One way could be turning the `generateCSIDriver` function into something like:

```
func (c *ManilaController) generateCSIDriver() *storagev1.CSIDriver {

	sch := runtime.NewScheme()
	_ = scheme.AddToScheme(sch)
	_ = apiextv1beta1.AddToScheme(sch)
	decode := serializer.NewCodecFactory(sch).UniversalDeserializer().Decode
	stream, _ := ioutil.ReadFile("assets/csidriver.yaml")

	obj, _, _ := decode(stream, nil, nil)

	csi_obj := obj.(*storagev1.CSIDriver)

	return csi_obj
```
By doing this, we're getting the static asset and decoding it into a `storagev1.CSIDriver` object, so we keep the static asset definition, and patching it according to the `env` passed to the operator.
wdyt

Reply:
There already is [ReadFile](https://github.com/openshift/csi-driver-manila-operator/blob/master/assets/assets.go#L11) that returns `[]byte` with a file from `assets/` and [ReadCSIDriverV1OrDie]( https://github.com/openshift/library-go/blob/3c66b317b110fe1a94bb0c9a9fa9f7a46df29941/pkg/operator/resource/resourceread/storage.go#L37) that parses `[]byte` it into a CSIDriver object.

And finally, there is [ApplyCSIDriver](https://github.com/openshift/library-go/blob/3c66b317b110fe1a94bb0c9a9fa9f7a46df29941/pkg/operator/resource/resourceapply/storage.go#L108), that creates or updates (re-creates) the CSI driver.

So this PR can be just a new function `syncCSIDriver` with few lines of code (+unit tests, that may be harder).

Reply:
Nice thank you for the hint, I wasn't aware about these functions.
Let me simplify the PR further, as we can just have a syncCSIDriver() that relies on the underlying functions!

Thanks!

Reply:
Hey @jsafrane, thanks for all your input so far, I managed to rely on the existing code as much as possible to sync the `CSIDriver` object and make the `FsGroupPolicy` a parameter that can be changed according to the env defined in the operator manifest.
As per your previous comment, this PR is now something that introduces the `syncCSIDriver` function that creates/updates the object according to the operator input. 

Here the main highlights:

* The CSIDriver resource is static and defined in `assets/`: the operator reads the file and build the object in case it doesn't exist, otherwise it applies it with the `CSIDriverSpec` passed as input;

* The `ApplyCSIDriver` function, provided in library-go, is used to update the immutable fields spec through deletion and recreation;

I guess we can now resolve this conversation, unless there's something I've overlooked.